src/streamed_orderbook.ts

```diff
@@ -0,0 +1,267 @@
+          )
+          break
+        case 'OrderCancellation':
+          this.account(data.owner).orders[parseInt(data.id)]!.validUntil = null
+          break
+        case 'OrderDeletion':
+          this.account(data.owner).orders[parseInt(data.id)] = null
+          break
+        case 'OrderPlacement':
+          this.account(data.owner).orders.push({
+            buyToken: data.buyToken,
+            sellToken: data.sellToken,
+            validFrom: parseInt(data.validFrom),
+            validUntil: parseInt(data.validUntil),
+            priceNumerator: BigInt(data.priceNumerator),
+            priceDenominator: BigInt(data.priceDenominator),
+            remainingAmount: BigInt(data.priceDenominator),
+          })
+          break
+        case 'SolutionSubmission':
+          if (!data.ext.reverted) {
+            this.updateBalance(
+              data.submitter,
+              '0',
+              amount => amount + BigInt(data.burntFees),
+            )
+            for (let i = 0; i < data.prices.length; i++) {
+              this.state.prices[data.tokenIdsForPrice[i]] = BigInt(data.prices[i])
+            }
+          }
+          break
+        case 'TokenListing':
+          this.state.tokens[data.id] = data.token
+          this.state.tokens[data.token] = data.id
+          break
+        case 'Trade':
+          if (!data.ext.reverted) {
+            this.updateBalance(
+              data.owner,
+              data.sellToken,
+              amount => amount - BigInt(data.executedSellAmount),
+            )
+            this.account(data.owner)
+              .orders[parseInt(data.orderId)]!
+              .remainingAmount -= BigInt(data.executedSellAmount)
+            this.updateBalance(
+              data.owner,
+              data.buyToken,
+              amount => amount + BigInt(data.executedBuyAmount),
+            )
+          }
+          break
+        case 'TradeReversion':
+          break
+        case 'WithdrawRequest':
+          this.updateWithdraw(
+            data.user,
+            data.token,
+            _ => BigInt(data.amount),
+          )
+          break
+        case 'Withdraw':
+          this.updateBalance(
+            data.user,
+            data.token,
+            amount => amount - BigInt(data.amount),
+          )
+          this.updateWithdraw(
+            data.user,
+            data.token,
+            _ => BigInt(0),
+          )
+          break
+        default:
+          console.error(ev)
+          throw new Error('unhandled event')
+      }
+    }
+  }
+
+  private markReverted(events: EventData[]) {
+    let revertedTradeCount = 0
+    for (let i = events.length - 1; i >= 0; i--) {
+      const ev = events[i]
+      const data = ev.returnValues
+      switch (ev.event) {
+        case 'Trade':
+          data.ext = { reverted: revertedTradeCount > 0 }
+          if (revertedTradeCount > 0) {
+            revertedTradeCount--
+          }
```

Comment:
Is this necessary? I was under the impression, that once a tradeReversion event was recorded all prior trade events will have been reverted too. Maybe we should turn this assumption into an assertion?

Reply:
Yes this is needed, since there are a total X trade reversions and X+Y trades, where Y is the number of trades in the winning latest solution.

This code has been removed since this `markReverted` method was confusing and not actually needed.